llama_index/query_engine/citation_query_engine.py

```diff
@@ -0,0 +1,299 @@
+        optimizer: Optional[BaseTokenUsageOptimizer] = None,
+        # class-specific args
+        **kwargs: Any,
+    ) -> "CitaitonQueryEngine":
+        """Initialize a CitaitonQueryEngine object."
+
+        Args:
+            retriever (BaseRetriever): A retriever object.
+            service_context (Optional[ServiceContext]): A ServiceContext object.
+            node_postprocessors (Optional[List[BaseNodePostprocessor]]): A list of
+                node postprocessors.
+            verbose (bool): Whether to print out debug info.
+            response_mode (ResponseMode): A ResponseMode object.
+            text_qa_template (Optional[QuestionAnswerPrompt]): A QuestionAnswerPrompt
+                object.
+            refine_template (Optional[RefinePrompt]): A RefinePrompt object.
+            simple_template (Optional[SimpleInputPrompt]): A SimpleInputPrompt object.
+            response_kwargs (Optional[Dict]): A dict of response kwargs.
+            use_async (bool): Whether to use async.
+            streaming (bool): Whether to use streaming.
+            optimizer (Optional[BaseTokenUsageOptimizer]): A BaseTokenUsageOptimizer
+                object.
+
+        """
+        retriever = retriever or index.as_retriever(**kwargs)
+
+        response_synthesizer = ResponseSynthesizer.from_args(
+            service_context=index.service_context,
+            text_qa_template=text_qa_template,
+            refine_template=refine_template,
+            response_mode=response_mode,
+            response_kwargs=response_kwargs,
+            use_async=use_async,
+            streaming=streaming,
+            optimizer=optimizer,
+            node_postprocessors=node_postprocessors,
+            verbose=verbose,
+        )
+
+        return cls(
+            retriever=retriever,
+            response_synthesizer=response_synthesizer,
+            callback_manager=index.service_context.callback_manager,
+            source_chunk_size=source_chunk_size,
+            source_chunk_overlap=source_chunk_overlap,
+            text_splitter=text_splitter,
+        )
+
+    def _create_citation_nodes(self, nodes: List[NodeWithScore]) -> List[NodeWithScore]:
+        """Modify retrieved nodes to be granular sources."""
+        text_splitter = TokenTextSplitter(chunk_size=256, chunk_overlap=20)
+
+        new_nodes: List[NodeWithScore] = []
+        for node in nodes:
+            splits = text_splitter.split_text_with_overlaps(node.node.get_text())
+
+            start_offset = 0
+            if node.node.node_info:
+                start_offset = int(node.node.node_info.get("start", 0))
+
+            for split in splits:
+                text = f"Source {len(new_nodes)+1}:\n{split.text_chunk}\n"
+
+                # NOTE currently this does not take into account escaped chars
+                num_char_overlap = split.num_char_overlap or 0
+                chunk_len = len(split.text_chunk)
+                new_node_info = {
```

Comment:
Not sure what else we put in node_info right now. But this would ignore any existing info there. I wonder if it's better to update instead of replace?

Reply:
Also, for our future data model. We should have an explicit field(s) for reference related information, rather than a catch all dict.

Reply:
Good point, will update instead. 

And I agree. I want to research some document models a bit and bring some ideas to SF next week